CameraX/app/src/main/java/github/nisrulz/example/cameraX/MainActivity.kt

```diff
@@ -0,0 +1,229 @@
+package github.nisrulz.example.cameraX
+
+import android.Manifest
+import android.content.ContentValues
+import android.content.pm.PackageManager
+import android.os.Build
+import android.os.Bundle
+import android.provider.MediaStore
+import android.util.Log
+import android.widget.Toast
+import androidx.appcompat.app.AppCompatActivity
+import androidx.camera.core.CameraSelector
+import androidx.camera.core.ImageCapture
+import androidx.camera.core.ImageCaptureException
+import androidx.camera.core.Preview
+import androidx.camera.lifecycle.ProcessCameraProvider
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import androidx.core.view.WindowCompat
+import androidx.core.view.WindowInsetsCompat
+import androidx.core.view.WindowInsetsControllerCompat
+import com.google.common.util.concurrent.ListenableFuture
+import github.nisrulz.example.cameraX.databinding.ActivityMainBinding
+import java.text.SimpleDateFormat
+import java.util.*
+
+private const val IMMERSIVE_FLAG_TIMEOUT = 500L
+
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var binding: ActivityMainBinding
+    private lateinit var imageCapture: ImageCapture
+
+    companion object {
+        private const val FILENAME_FORMAT = "yyyy-MM-dd-HH-mm-ss-SSS"
+
+        private const val REQUEST_CODE_PERMISSIONS = 10
+
+        private val REQUIRED_PERMISSIONS = arrayOf(Manifest.permission.CAMERA)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityMainBinding.inflate(layoutInflater)
+        binding.apply {
+            setContentView(root)
+        }
+
+        if (allPermissionsGranted()) {
+            startCamera()
+        } else {
+            // Request for permissions mentioned in [REQUIRED_PERMISSIONS].
+            // This will invoke the [onRequestPermissionsResult]
+            ActivityCompat.requestPermissions(
+                this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS
+            )
+        }
```

Comment:
As per the official document quote
>Ask for permissions in context, when the user starts to interact with the feature that requires it.

I am thinking to execute this only when a user tries to click on the capture button. 
WDYT?